akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -22,6 +25,85 @@ object ORMap {
    */
   def unapply[A, B <: ReplicatedData](m: ORMap[A, B]): Option[Map[A, B]] = Some(m.entries)
 
+  sealed trait DeltaOp extends ReplicatedDelta with RequiresCausalDeliveryOfDeltas {
+    type T = DeltaOp
+    override def zero: DeltaReplicatedData
+  }
+
+  private[akka] final val VanillaORMapTag = 0
+  private[akka] final val PNCounterMapTag = 1
+  private[akka] final val ORMultiMapTag = 2
+  private[akka] final val ORMultiMapDeltaDeltaTag = 3
+  private[akka] final val LWWMapTag = 4
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def tagDelta(op: DeltaOp)(tag: Int) = op match {
+    case p: PutDeltaOp[_, _]       ⇒ p.copy(zeroTag = tag)
```

Comment:
I think we can avoid these matches, by adding
`def withZeroTag(tag: Int): DeltaOp` in `DeltaOp` and implement it in the concrete types, e.g. UpdateDeltaOp:

    override def withZeroTag(tag: Int): UpdateDeltaOp = copy(zeroTag = tag)

but I'm also thinking of why the tagging is done afterwards. Couldn't `ORMap` have a constructor parameter with the `zeroTag`, e.g. `PNCounterMap` knows it's tag when it creates the `underlying: ORMap`. Then `ORMap` could use the right tag immediately when creating the `DeltaOp` instances?